File: Solutions/938_Range_Sum_of_BST/938_Range_Sum_of_BST.rb
def range_sum_bst(root, low, high)
  stack = [root]
  result = 0
  
  until stack.empty?
    node = stack.pop
    
    result += node.val if (low..high).cover?(node.val)
   
    stack << node.left if node.left && node.val >= low
    stack << node.right if node.right && node.val <= high
  end
  
  result
end

#Third fastest
def range_sum_bst(root, l, r)
    searcher = Searcher.new
    searcher.dfs(root, l, r)
    searcher.sum
end

class Searcher
    def initialize
        @sum = 0
    end
    attr_reader :sum
    
    def dfs(root, l, r)
        return if root.nil?

        @sum += root.val if root.val.between?(l, r)

        if root.val >= l
            dfs(root.left, l, r)
        end

        if root.val <= r
            dfs(root.right, l, r)
        end
    end
end


#Second fastest
def range_sum_bst(node, l, r, sum = 0)
    return 0 unless node 
    
    sum += node.val if node.val >= l && node.val <= r
    sum = range_sum_bst(node.left, l, r, sum) if node.left
    sum = range_sum_bst(node.right, l, r, sum) if node.right
    
    return sum
end


#Fastest
def range_sum_bst(node, l, r, sum = 0)
    return 0 unless node 
    
    if node.val < l
        sum = range_sum_bst(node.right, l, r, sum)
    elsif node.val > r
        sum = range_sum_bst(node.left, l, r, sum)
    else
        sum += node.val + range_sum_bst(node.right, l, r, sum) + range_sum_bst(node.left, l, r, sum)
    end
    
    return sum
end

File: Solutions/938_Range_Sum_of_BST/938_Range_Sum_of_BST.js
//Fourth fastest
var rangeSumBST = function(root, L, R) {
    let sum = 0;
    function calSum(node) {
        if(node != 0 && node.val >= L && node.val <=R){
            sum+= node.val
        }
        if( node.val <= R && node.right != null)
        calSum(node.right);
        if(node.val >= L && node.left != null)
        calSum(node.left)
    }
    calSum(root);
    return sum;
    
};

//Third fastest
var rangeSumBST = function(root, L, R) {
    let sum = 0;
     traverse(root);
     return sum;
     
     function traverse(node) {
         
         if(!node) {
             return;
         }
         
         if(node.val < L) {
             traverse(node.right);
             return;
         }
         
         if(node.val > R) {
             traverse(node.left);
             return;
         }
         
         sum+= node.val;
         traverse(node.right)
         traverse(node.left)
         
         
     }
 };
 
 //need to traverse the bst
 
 //check if the node is greater =  than L
 //check if the node is less or = to R
 // add to L

//Second fastest
var rangeSumBST = function(root, L, R) {
    // travere by inorder -> sorted result
    let stack = [root];
    let result = [];
    while (stack.length) {
      let cur = stack.pop();
      result.push(cur.val);
      cur.left && stack.push(cur.left);
      cur.right && stack.push(cur.right);
    }
    return result.reduce((acc, cur) => {
      if (L <= cur && cur <= R) {
        return acc + cur;
      }
      return acc;
    }, 0);  
  };

//Fastest
var rangeSumBST = function(root, L, R) {
    let total=0
    const search=(node)=>{
    if (!node)return 0
    if(node.val>=L){
       if(node.val<=R){
           total+=node.val
       }
    }
   node.val>L && search(node.left)
   node.val<R && search(node.right)
    }
    search(root)
    return total
}